digital_land/utils/add_endpoints_utils.py

```diff
@@ -0,0 +1,708 @@
+from digital_land.phase.migrate import MigratePhase
+from digital_land.phase.normalise import NormalisePhase
+from digital_land.phase.organisation import OrganisationPhase
+from digital_land.phase.parse import ParsePhase
+from digital_land.phase.patch import PatchPhase
+from digital_land.phase.phase import Phase
+from digital_land.phase.prefix import EntityPrefixPhase
+from digital_land.phase.prune import FieldPrunePhase
+from digital_land.phase.reference import EntityReferencePhase
+
+from digital_land.pipeline import run_pipeline
+from digital_land.organisation import Organisation
+from digital_land.schema import Schema
+from digital_land.store.item import ItemStore
+from digital_land.update import add_source_endpoint
+from digital_land.register import hash_value
+from pathlib import Path
+from urllib.error import URLError
+
+
+# =====================================================================
+# Main Process Tasks
+# =====================================================================
+def task_preprocess(ctx):
+    """
+    preparatory steps to tidy up previous runs,
+    and populate the context
+    :param ctx:
+    :return:
+    """
+
+    collection_dir = ctx.obj["COLLECTION_DIR"]
+    root_coll_dir = collection_dir.parent
+    datasource_log_dir = root_coll_dir / "log"
+    issues_log_dir = root_coll_dir / "log"
+    log_dir = collection_dir / "log"
+    organisation_dir = root_coll_dir / "organisation"
+    tmp_dir = root_coll_dir / "tmp"
+    collection_resource_dir = collection_dir / "resource"
+
+    ctx.obj["PIPELINE_DIR"] = Path(ctx.obj["PIPELINE"].path)
+    ctx.obj["DATASOURCE_LOG_DIR"] = datasource_log_dir
+    ctx.obj["ISSUES_LOG_DIR"] = issues_log_dir
+    ctx.obj["LOG_DIR"] = log_dir
+    ctx.obj["ORGANISATION_DIR"] = organisation_dir
+    ctx.obj["TMP_DIR"] = tmp_dir
+    ctx.obj["COLLECTION_RESOURCE_DIR"] = collection_resource_dir
+
+    # clean up previous run data
+    preprocess_collection_csvs(ctx)
+
+    if log_dir.is_dir():
+        shutil.rmtree(log_dir)
+    if organisation_dir.is_dir():
+        shutil.rmtree(organisation_dir)
+
+    try:
+        os.mkdir(log_dir)
+        os.mkdir(organisation_dir)
+        os.mkdir(tmp_dir)
+    except OSError:
+        pass
+
+
+def task_create_source_and_endpoint_entries(ctx):
+    """
+    appends entries to both source.csv and endpoints.csv from
+    the records in the csv at csv_file_path
+    :param ctx:
+    :return:
+    """
+
+    # read and process each record in the csv at csv_file_path
+    csv_file_path = ctx.obj["CSV_FILE_PATH"]
+
+    cols = [
+        "dataset",
+        "organisation",
+        "name",
+        "documentation-url",
+        "endpoint-url",
+        "start-date",
+    ]
+    csv_file_df = pd.read_csv(
+        csv_file_path,
+        header=0,
+        index_col=False,
+        usecols=cols,
+    )
+
+    collection = Collection()
+    ctx.obj["COLLECTION"] = collection
+
+    for idx, row in csv_file_df.iterrows():
```

Comment:
should it fail if any of the endpoints fail?

Reply:
I think we need to add in more standard validation for endpoints and sources. perhaps we can leverage the CsvStore class to create reusable functions for this. It can be a static method attached to one of the classes.

The aim of this is to make it available while using the class (we can include the validations more frequently) This will improve the code base by ensuring that the same methods are used to validate the values.